src/equations/hyperbolic_serre_green_naghdi_1d.jl

```diff
@@ -0,0 +1,364 @@
+
+        # Plain: h_t + (h v)_x = 0
+        #
+        # Split form for energy conservation:
+        # h_t + h_x v + h v_x = 0
+        @. dh = -(h_x * v + h * v_x)
+
+        # Plain: h v_t + h v v_x + g (h + b) h_x
+        #              + ... = 0
+        #
+        # Split form for energy conservation:
+        # h v_t + g (h (h + b))_x - g (h + b) h_x
+        #       + 1/2 h (v^2)_x - 1/2 v^2 h_x  + 1/2 v (h v)_x - 1/2 h v v_x
+        #       + λ/6 η^2 / h^2 h_x + λ/3 η_x - λ/3 η/h η_x - λ/6 (η^2 / h)_x
+        #       + λ/2 b_x - λ/2 η/h b_x = 0
+        λ_6 = λ / 6
+        λ_3 = λ / 3
+        λ_2 = λ / 2
+        @. dv = -(g * h_hpb_x - g * (h + b) * h_x
+                    + 0.5 * h * v2_x - 0.5 * v^2 * h_x
+                    + 0.5 * hv_x * v - 0.5 * h * v * v_x
+                    + λ_6 * (η_over_h * η_over_h * h_x - η2_h_x)
+                    + λ_3 * (1 - η_over_h) * η_x
+                    + λ_2 * (1 - η_over_h) * b_x) / h
+
+        # Plain: h w_t + h v w_x = λ - λ η / h
+        #
+        # Split form for energy conservation:
+        # h w_t + 1/2 (h v w)_x + 1/2 h v w_x
+        #       - 1/2 h_x v w - 1/2 h w v_x = λ - λ η / h
+        @. dw = ( -(  0.5 * hvw_x
+                    + 0.5 * h * v * w_x
+                    - 0.5 * h_x * v * w
+                    - 0.5 * h * w * v_x) + λ * (1 - η_over_h)) / h
+
+        # No special split form for energy conservation required:
+        # η_t + v η_x + 3/2 v b_x = w
+        @. dη = -v * η_x - 1.5 * v * b_x + w
+    end
+
+    @trixi_timeit timer() "source terms" calc_sources!(dq, q, t, source_terms, equations,
+                                                       solver)
+
+    return nothing
+end
+
+@inline function prim2cons(q, equations::HyperbolicSerreGreenNaghdiEquations1D)
+    h = waterheight(q, equations)
+    v = velocity(q, equations)
+    b = bathymetry(q, equations)
+
+    hv = h * v
+    hw = h * q[4]
+    hη = h * q[5]
+    return SVector(h, hv, b, hw, hη)
+end
+
+@inline function cons2prim(u, equations::HyperbolicSerreGreenNaghdiEquations1D)
+    h, hv, b = u
+
+    eta = h + b
+    v = hv / h
+    D = equations.eta0 - b
+    w = u[4] / h
+    η = u[5] / h
+    return SVector(eta, v, D, w, η)
+end
+
+# The entropy/energy takes the whole `q` for every point in space
+"""
+    energy_total_modified(q_global, equations::HyperbolicSerreGreenNaghdiEquations1D, cache)
+
+Return the modified total energy of the primitive variables `q_global` for the
+[`HyperbolicSerreGreenNaghdiEquations1D`](@ref).
+It contains additional terms compared to the usual [`energy_total`](@ref)
+modeling non-hydrostatic contributions. The `energy_total_modified`
+is a conserved quantity (for periodic boundary conditions).
+
+For a [`bathymetry_mild_slope`](@ref) (and a [`bathymetry_flat`](@ref)),
+the total modified energy is given by
+```math
+\\frac{1}{2} g \\eta^2 + \\frac{1}{2} h v^2 +
+\\frac{1}{6} h w^2 + \\frac{\\lambda}{6} h (1 - \\eta / h)^2.
+```
```

Comment:
In this case, it's not strictly required to get the full vector

Reply:
Ok, I think that's fine. See also my suggestion below.